In [2]:
import pandas as pd
import numpy as np
import re

print("✅ Jupyter is working correctly!")



✅ Jupyter is working correctly!


In [3]:
df = pd.read_csv("../data/netflix_titles.csv")


In [4]:
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [5]:
df['date_added'].head()

0    September 25, 2021
1    September 24, 2021
2    September 24, 2021
3    September 24, 2021
4    September 24, 2021
Name: date_added, dtype: object

In [6]:
df.shape

(8807, 12)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [8]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [9]:
df.describe(include = 'all')

show_id   type                 title       director  \
count     8807   8807                  8807           6173   
unique    8807      2                  8807           4528   
top         s1  Movie  Dick Johnson Is Dead  Rajiv Chilaka   
freq         1   6131                     1             19   
mean       NaN    NaN                   NaN            NaN   
std        NaN    NaN                   NaN            NaN   
min        NaN    NaN                   NaN            NaN   
25%        NaN    NaN                   NaN            NaN   
50%        NaN    NaN                   NaN            NaN   
75%        NaN    NaN                   NaN            NaN   
max        NaN    NaN                   NaN            NaN   

                      cast        country       date_added  release_year  \
count                 7982           7976             8797   8807.000000   
unique                7692            748             1767           NaN   
top     David Attenborough  United States  January 1, 2020           NaN   
freq                    19           2818              109           NaN   
mean                   NaN            NaN              NaN   2014.180198   
std                    NaN            NaN              NaN      8.819312   
min                    NaN            NaN              NaN   1925.000000   
25%                    NaN            NaN              NaN   2013.000000   
50%                    NaN            NaN              NaN   2017.000000   
75%                    NaN            NaN              NaN   2019.000000   
max                    NaN            NaN              NaN   2021.000000   

       rating  duration                     listed_in  \
count    8803      8804                          8807   
unique     17       220                           514   
top     TV-MA  1 Season  Dramas, International Movies   
freq     3207      1793                           362   
mean      NaN       NaN                           NaN   
std       NaN       NaN                           NaN   
min       NaN       NaN                           NaN   
25%       NaN       NaN                           NaN   
50%       NaN       NaN                           NaN   
75%       NaN       NaN                           NaN   
max       NaN       NaN                           NaN   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN


### 🧠 Observations from Dataset Overview

1. **Shape of Dataset:**  
   - Total **8807 rows and 12 columns** — a moderate-sized dataset.

2. **Data Types Summary:**  
   - Includes both **categorical** (`type`, `title`, `director`, `country`, `rating`, `listed_in`, etc.) and **numeric** (`release_year`) features.  
   - `release_year` is clean and numeric — useful for time-based trend analysis.

3. **Unique Value Insights:**  
   - `type` → 2 unique values → **Movies** and **TV Shows**  
   - `country` → 748 unique countries → shows global diversity  
   - `rating` → 17 unique content ratings (e.g., TV-MA, PG, R)  
   - `listed_in` → 514 unique genre/category combinations — wide content variety  

4. **Most Frequent (Top) Values:**  
   - **Type:** Movie (6,131 entries)  
   - **Country:** United States (2,818 titles)  
   - **Rating:** TV-MA (3,207 entries) → mostly adult-oriented content  
   - **Director:** Rajiv Chilaka (19 titles)  
   - **Cast:** David Attenborough (19 titles)

5. **Missing Values (indirectly seen):**  
   - Columns like `director`, `cast`, and `country` have missing entries.  
   - Example: `director` count = 6,173 → roughly **30% missing** values.  
   - These will need cleaning or imputation later.

6. **Release Year Summary:**  
   - Range: **1925 – 2021** (≈96 years)  
   - Mean ≈ **2014**, showing most Netflix content is modern.

7. **Duration Field:**  
   - 220 unique values, mixed with *minutes* and *seasons*.  
   - Will need to separate **Movies vs TV Shows** and clean this field.

8. **Description:**  
   - No missing values → could be used later for **text/NLP-based analysis**.

---

### 📋 Early Insights / Next Steps
- Dataset is **movie-dominant** and focuses mostly on **modern (post-2010) content**.  
- **Missing values** in key columns — will address in data cleaning phase.  
- Need to split analysis by **Type** (`Movie` vs `TV Show`).  
- `listed_in` column will help in exploring **popular genres/categories** by country or type.

---

🧩 *Next Step:* Move to **Phase 3 — Data Cleaning & Preparation**,  
where we’ll handle nulls, duplicates, and fix inconsistent fields like `duration`.


In [10]:
print(f"Before removing duplicates: {df.shape}")
df.drop_duplicates(inplace=True)
print(f"After removing duplicates: {df.shape}")


Before removing duplicates: (8807, 12)
After removing duplicates: (8807, 12)


In [11]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [12]:
for col in ['rating', 'duration', 'director', 'cast', 'country']:
    df[col] = df[col].fillna("unknown")
df['date_added'] = df['date_added'].fillna(df['date_added'].mode()[0])



In [13]:
df.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [14]:
print("Initial dtype:", df['date_added'].dtype)
print("Sample values:")
display(df['date_added'].head(10))

Initial dtype: object
Sample values:


0    September 25, 2021
1    September 24, 2021
2    September 24, 2021
3    September 24, 2021
4    September 24, 2021
5    September 24, 2021
6    September 24, 2021
7    September 24, 2021
8    September 24, 2021
9    September 24, 2021
Name: date_added, dtype: object

In [15]:
df['date_added_raw'] = df['date_added']
df['date_added'] = df['date_added'].astype(str).str.strip()
df.loc[df['date_added'].isin(['', 'nan', 'None', 'NaN', 'NoneType']), 'date_added'] = np.nan

In [16]:
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

In [17]:
num_failed = df['date_added_parsed'].isna().sum()
print(f"Parsed OK: {df.shape[0] - num_failed}, Failed/NaT: {num_failed}")

KeyError: 'date_added_parsed'

In [18]:
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')
df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month

print("Final dtype:", df['date_added'].dtype)
print("Number of NaT in date_added:", df['date_added'].isna().sum())
df[['title', 'date_added', 'year_added', 'month_added']]

Final dtype: datetime64[ns]
Number of NaT in date_added: 0


title date_added  year_added  month_added
0      Dick Johnson Is Dead 2021-09-25        2021            9
1             Blood & Water 2021-09-24        2021            9
2                 Ganglands 2021-09-24        2021            9
3     Jailbirds New Orleans 2021-09-24        2021            9
4              Kota Factory 2021-09-24        2021            9
...                     ...        ...         ...          ...
8802                 Zodiac 2019-11-20        2019           11
8803            Zombie Dumb 2019-07-01        2019            7
8804             Zombieland 2019-11-01        2019           11
8805                   Zoom 2020-01-11        2020            1
8806                 Zubaan 2019-03-02        2019            3

[8807 rows x 4 columns]

In [19]:
text_cols = ['title', 'director', 'cast', 'country', 'rating', 'listed_in']

for col in text_cols:
  df[col] = df[col].astype(str).str.strip().str.title()

In [20]:
df['duration_int'] = df['duration'].str.extract('(\d+)').astype(float)
df['duration_unit'] = df['duration'].str.extract('([A-Za-z]+)').astype(str)



<>:1: SyntaxWarning: invalid escape sequence '\d'
<>:1: SyntaxWarning: invalid escape sequence '\d'
C:\Users\BHANARAM\AppData\Local\Temp\ipykernel_14224\3734778939.py:1: SyntaxWarning: invalid escape sequence '\d'
  df['duration_int'] = df['duration'].str.extract('(\d+)').astype(float)


In [21]:
df.info()
df.sample(5)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   show_id         8807 non-null   object        
 1   type            8807 non-null   object        
 2   title           8807 non-null   object        
 3   director        8807 non-null   object        
 4   cast            8807 non-null   object        
 5   country         8807 non-null   object        
 6   date_added      8807 non-null   datetime64[ns]
 7   release_year    8807 non-null   int64         
 8   rating          8807 non-null   object        
 9   duration        8807 non-null   object        
 10  listed_in       8807 non-null   object        
 11  description     8807 non-null   object        
 12  date_added_raw  8807 non-null   object        
 13  year_added      8807 non-null   int32         
 14  month_added     8807 non-null   int32         
 15  dura

show_id     type                      title       director  \
3327   s3328  TV Show            A Year In Space        Unknown   
6245   s6246    Movie       Barrio Universitario  Esteban Vidal   
1806   s1807    Movie                    Cadaver  Jarand Herdal   
7839   s7840    Movie      Rebel Without A Cause   Nicholas Ray   
1468   s1469    Movie  What Happened To Mr. Cha?   Kim Dong-Kyu   

                                                   cast        country  \
3327                                        Scott Kelly  United States   
6245  Fabrizio Copano, Rodrigo Salinas, Pedro Rumino...          Chile   
1806  Gitte Witt, Thomas Gullestad, Thorbjørn Harr, ...         Norway   
7839  James Dean, Natalie Wood, Sal Mineo, Jim Backu...  United States   
1468           Cha In-Pyo, Cho Dal-Hwan, Song Jae-Ryong    South Korea   

     date_added  release_year rating  duration  \
3327 2019-11-01          2015  Tv-Pg  1 Season   
6245 2018-08-19          2013  Tv-14    91 min   
1806 2020-10-22          2020  Tv-Ma    87 min   
7839 2019-11-01          1955  Pg-13   111 min   
1468 2021-01-01          2021  Tv-Ma   102 min   

                                              listed_in  \
3327  Docuseries, International Tv Shows, Science & ...   
6245                     Comedies, International Movies   
1806                Horror Movies, International Movies   
7839                             Classic Movies, Dramas   
1468                     Comedies, International Movies   

                                            description    date_added_raw  \
3327  Two astronauts attempt to brave a life in Eart...  November 1, 2019   
6245  Four friends from a humble technical school co...   August 19, 2018   
1806  When a peculiar hotel dinner show draws crowds...  October 22, 2020   
7839  When Jim Stark arrives in a new town with his ...  November 1, 2019   
1468  With the peak of his career long behind him, a...   January 1, 2021   

      year_added  month_added  duration_int duration_unit  
3327        2019           11           1.0        Season  
6245        2018            8          91.0           min  
1806        2020           10          87.0           min  
7839        2019           11         111.0           min  
1468        2021            1         102.0           min

In [22]:
from datetime import datetime

df['content_age'] = datetime.now().year - df['release_year']
df['is_modern'] = df['release_year'] >= 2015
df['num_genres'] = df['listed_in'].apply(lambda x: len(x.split(',')))
df['has_known_director'] = df['director'] != 'Unknown'


In [23]:
df.to_csv("../data/netflix_cleaned.csv", index=False)
print("✅ Cleaned dataset saved successfully!")


✅ Cleaned dataset saved successfully!


In [24]:
 df['duration_int']

0        90.0
1         2.0
2         1.0
3         1.0
4         2.0
        ...  
8802    158.0
8803      2.0
8804     88.0
8805     88.0
8806    111.0
Name: duration_int, Length: 8807, dtype: float64